src/LandIce/evaluators/LandIce_BasalMeltRate_Def.hpp

```diff
@@ -107,55 +120,173 @@ postRegistrationSetup(typename Traits::SetupData d, PHX::FieldManager<Traits>& f
   if (d.memoizer_active()) memoizer.enable_memoizer();
 }
 
+template<typename EvalT, typename Traits, typename VelocityST, typename MeltEnthST>
+KOKKOS_INLINE_FUNCTION
+void BasalMeltRate<EvalT,Traits,VelocityST,MeltEnthST>::
+operator() (const Basal_Melt_Rate_Collapsed_Tag& tag, const int& sideSet_idx) const {
+
+  const int dim = nodal ? numSideNodes : numSideQPs;
+
+  // Get the local data of side and cell
+  const int cell = sideSet.elem_LID(sideSet_idx);
+  const int side = sideSet.side_local_id(sideSet_idx);
+
+  for (int node = 0; node < dim; ++node) {
+    //always in presence of water on shelves (assuming that beta==0 <==> on shelves)
+    bool isThereWaterHere = isThereWater || (beta(sideSet_idx,node) == 0.0);
+    ScalarT diffEnthalpy = Enthalpy(sideSet_idx,node) - EnthalpyHs(sideSet_idx,node);
```

Comment:
Have you tried using `const auto` rather than `ScalarT` for (const) temporary expressions? I think it might make Sacado simply store the expression on the rhs, and Sacato will evaluate it once at the end (possibly doing some better optimization with temporaries when expanding the expression and evaluating it).

Also, have you tried `const auto&` when you just extract one entry? It might save some register usage on GPU.

Reply:
I think you mentioned this in the previous PR. I think I tried this a few years ago and it results in an error when using DFad and/or SFad and/or on GPU. It's worth trying again though but maybe we should wait until we make another pass at optimizing the code.

The benefit on GPU will depend on a few things. If the memory access is contiguous in global memory, it probably makes sense to not store the temp. If it's not, then it's probably better to store in local memory.